config/interpolate_walk.go

```diff
@@ -85,7 +89,13 @@ func (w *interpolationWalker) Map(m reflect.Value) error {
 func (w *interpolationWalker) MapElem(m, k, v reflect.Value) error {
 	w.csData = k
 	w.csKey = append(w.csKey, k)
-	w.key = append(w.key, k.String())
+
+	if w.sliceIndex != -1 {
+		w.key = append(w.key, fmt.Sprintf("%d.%s", w.sliceIndex, k.String()))
+	} else {
+		w.key = append(w.key, k.String())
+	}
```

Comment:
The reason it is like this is because of the lack of Enter and Exit for Map and Slice elements - meaning we have no opportunity to just mutate the key during slice visit (at least, not in a way that is clean). Not really sure how easy it is to test - however, there is test coverage that this is correct via the next layer up.